subsys/net/ip/connection.c

```diff
@@ -576,94 +576,283 @@ static bool conn_are_endpoints_valid(struct net_pkt *pkt, uint8_t family,
+		goto out;
 	}
 
-	if (!(conn->flags & NET_CONN_LOCAL_ADDR_SET)) {
-		return NET_CONTINUE;
+	if (conn->cb(conn, raw_pkt, NULL, NULL, conn->user_data) == NET_DROP) {
+		net_pkt_unref(raw_pkt);
 	}
 
-	struct net_if *pkt_iface = net_pkt_iface(pkt);
-	struct sockaddr_ll *local;
-	struct net_pkt *raw_pkt;
+out:
+	net_pkt_cursor_restore(pkt, &cur);
+}
+#endif /* defined(CONFIG_NET_SOCKETS_PACKET) || defined(CONFIG_NET_SOCKETS_INET_RAW) */
 
-	local = (struct sockaddr_ll *)&conn->local_addr;
+#if defined(CONFIG_NET_SOCKETS_PACKET)
+enum net_verdict net_conn_packet_input(struct net_pkt *pkt, uint16_t proto)
+{
+	bool raw_pkt_delivered = false;
+	bool raw_pkt_continue = false;
+	struct sockaddr_ll *local;
+	struct net_conn *conn;
 
-	if (local->sll_ifindex != net_if_get_by_iface(pkt_iface)) {
+	/* Only accept input with AF_PACKET family. */
+	if (net_pkt_family(pkt) != AF_PACKET) {
 		return NET_CONTINUE;
 	}
 
-	NET_DBG("[%p] raw match found cb %p ud %p", conn, conn->cb,
-		conn->user_data);
+	NET_DBG("Check proto 0x%04x listener for pkt %p family %d",
+		proto, pkt, net_pkt_family(pkt));
 
-	raw_pkt = net_pkt_clone(pkt, K_MSEC(CONFIG_NET_CONN_PACKET_CLONE_TIMEOUT));
-	if (!raw_pkt) {
-		net_stats_update_per_proto_drop(pkt_iface, proto);
-		NET_WARN("pkt cloning failed, pkt %p dropped", pkt);
-		return NET_DROP;
+	k_mutex_lock(&conn_lock, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&conn_used, conn, node) {
+		if (!is_iface_matching(conn, pkt)) {
+			continue; /* wrong interface */
+		}
+
+		/* If there are other listening connections than
+		 * AF_PACKET, the packet shall be also passed back to
+		 * net_conn_input() in upper layer processing in order to
+		 * re-check if there is any listening socket interested
+		 * in this packet.
+		 */
+		if (conn->family != AF_PACKET) {
+			raw_pkt_continue = true;
+			continue; /* wrong family */
+		}
+
+		if (conn->type == SOCK_DGRAM && !net_pkt_is_l2_processed(pkt)) {
+			/* If DGRAM packet sockets are present, we shall continue
+			 * with this packet regardless the result.
+			 */
+			raw_pkt_continue = true;
+			continue; /* L2 not processed yet */
+		}
+
+		if (conn->type == SOCK_RAW && net_pkt_is_l2_processed(pkt)) {
+			continue; /* L2 already processed */
+		}
+
+		if (conn->proto == 0) {
+			continue; /* Local proto 0 doesn't forward any packets */
+		}
+
+		if (conn->proto != proto) {
+			/* Allow proto mismatch if socket was created with ETH_P_ALL, or it's raw
+			 * packet socket input (proto ETH_P_ALL).
+			 */
+			if (conn->proto != ETH_P_ALL && proto != ETH_P_ALL) {
+				continue; /* wrong protocol */
+			}
+		}
+
+		/* Apply protocol-specific matching criteria... */
+
+		if (!(conn->flags & NET_CONN_LOCAL_ADDR_SET)) {
+			continue;
+		}
+
+		local = (struct sockaddr_ll *)&conn->local_addr;
+		if (local->sll_ifindex != net_if_get_by_iface(net_pkt_iface(pkt))) {
+			continue;
+		}
+
+		conn_raw_socket_deliver(pkt, conn, false);
+
+		raw_pkt_delivered = true;
```

Comment:
Is it still considered `delivered` if `conn_raw_socket_deliver` failed? For example if the clone failed.

Reply:
I've renamed the variable to `raw_sock_found` to avoid confusion, it shouldn't really matter at this point if we've actually managed to pass the packet to the sock.